yt/yt/ytlib/api/native/client_queries.cpp

```diff
@@ -38,6 +40,305 @@ using namespace NYTree;
+    const std::optional<std::vector<TString>>& lookupKeys,
+    NTransactionClient::TTimestamp timestamp,
+    const NLogging::TLogger& logger)
+{
+    auto asyncActiveRecord = LookupQueryTrackerRecord<TActiveQueryDescriptor>(
+        queryId,
+        client,
+        root + "/active_queries",
+        "active",
+        lookupKeys,
+        timestamp);
+    auto asyncFinishedRecord = LookupQueryTrackerRecord<TFinishedQueryDescriptor>(
+        queryId,
+        client,
+        root + "/finished_queries",
+        "finished",
+        lookupKeys,
+        timestamp);
+
+    auto error = WaitFor(AnySucceeded(std::vector{asyncActiveRecord.AsVoid(), asyncFinishedRecord.AsVoid()}));
+    if (!error.IsOK()) {
+        THROW_ERROR_EXCEPTION(NQueryTrackerClient::EErrorCode::QueryNotFound,
+            "Query %v is not found neither in active nor in finished query tables",
+            queryId)
+            << error;
+    }
+    bool isActive = asyncActiveRecord.IsSet() && asyncActiveRecord.Get().IsOK();
+    bool isFinished = asyncFinishedRecord.IsSet() && asyncFinishedRecord.Get().IsOK();
+    YT_VERIFY(isActive || isFinished);
+    if (isActive && isFinished) {
+        const auto& Logger = logger;
+        YT_LOG_ALERT(
+            "Query is found in both active and finished query tables "
+            "(QueryId: %v, Timestamp: %v)",
+            queryId,
+            timestamp);
+    }
+    if (isActive) {
+        return PartialRecordToQuery(asyncActiveRecord.Get().Value());
+    } else {
+        return PartialRecordToQuery(asyncFinishedRecord.Get().Value());
+    }
+}
+
+void ValidateQueryPermissions(
+    TQueryId queryId,
+    const TString& root,
+    NTransactionClient::TTimestamp timestamp,
+    const TString& user,
+    const IClientPtr& client,
+    EPermission permission,
+    const NLogging::TLogger& logger)
+{
+    std::vector<TString> lookupKeys = {"user", "access_control_object"};
+    auto query = LookupQuery(queryId, client, root, lookupKeys, timestamp, logger);
+    if (CheckAccessControl(user, query.AccessControlObject, *query.User, client, permission) == NSecurityClient::ESecurityAction::Deny) {
+        ThrowAccessDeniedException(queryId, permission, user, query.AccessControlObject, *query.User);
+    }
+}
+
+std::vector<TString> GetAcosForSubjects(const THashSet<TString>& subjects, const IClientPtr& client)
+{
+    // Get all access control objects.
+    TGetNodeOptions options;
+    options.Attributes = {
+        "principal_acl",
+    };
+    options.ReadFrom = EMasterChannelKind::Cache;
+
+    auto allAcosOrError = WaitFor(client->GetNode("//sys/access_control_object_namespaces/queries", options));
+
+    if (!allAcosOrError.IsOK()) {
+        THROW_ERROR_EXCEPTION("Error while fetching all access control objects in the namespace \"queries\"")
+            << allAcosOrError;
+    }
+
+    auto allAcos = ConvertToNode(allAcosOrError.Value())->AsMap()->GetChildren();
+
+    std::vector<TString> acosForUser;
```

Comment:
Let's add `acosForUser.reserve(allAcos.size())` here to avoid extra allocations

Reply:
Added `reserve(10)` after discussion since growth of ACOs is linear, while ACOs for a single user expected to be constant on average